openvdb/points/PointMove.h

```diff
@@ -509,13 +535,41 @@ struct GlobalMovePointsOp
         IndexArray& offsets, const IndexTripleArray& indices) const
     {
         auto& targetArray = mTargetHandles.getArray(targetOffset);
+        targetArray.loadData();
+        targetArray.expand();
+
+        // build a sorted index vector that references the indices in order of their source
+        // leafs and voxels to ensure determinism in the resulting point orders
+
+        std::vector<int> sortedIndices(indices.size());
+        std::iota(std::begin(sortedIndices), std::end(sortedIndices), 0);
+        std::sort(std::begin(sortedIndices), std::end(sortedIndices),
+            [&](int i, int j)
+            {
+                const Index& indexI0(std::get<0>(indices[i]));
+                const Index& indexJ0(std::get<0>(indices[j]));
+                if (indexI0 < indexJ0)          return true;
+                if (indexI0 > indexJ0)          return false;
+                return std::get<2>(indices[i]) < std::get<2>(indices[j]);
+            }
+        );
+
+        for (const auto& index : sortedIndices) {
+            const auto& it = indices[index];
 
-        for (const auto& it : indices) {
             const auto& sourceArray = mSourceHandles.getConstArray(std::get<0>(it));
-            const Index tgtOffset = indexOffsetFromVoxel(std::get<1>(it), targetLeaf, offsets);
-            OPENVDB_NO_DEPRECATION_WARNING_BEGIN
-            targetArray.set(tgtOffset, sourceArray, std::get<2>(it));
-            OPENVDB_NO_DEPRECATION_WARNING_END
+
+            const Index sourceOffset = std::get<2>(it);
+            const Index targetOffset = indexOffsetFromVoxel(std::get<1>(it), targetLeaf, offsets);
+
+#if OPENVDB_ABI_VERSION_NUMBER >= 6
+            sourceArray.loadData();
+            SingleCopyIterator singleCopyIterator;
+            singleCopyIterator.setIndices(sourceOffset, targetOffset);
+            targetArray.copyValues(sourceArray, singleCopyIterator);
+#else
```

Comment:
There's a lot of unnecessary virtual methods being called here. copyValues calls loadData on both source and targte, and expand and compact on the target. You can alleviate the call to compact with the last argument but that still leaves the others. It seems that what you really want is to loop over all source arrays and load/expand the data once, then call copyValuesUnsafe() with range checking as true. I think I mentioned this in the original PR for this implementation.

Reply:
I'll look at doing what you're suggesting here. Bear in mind, the main purpose of this PR is to eliminate the use of a deprecated function in this (very rare) code branch for ABI=6+. The current ABI=5 implementation is a lot of code but essentially provides similar performance to the newer method. 

Ultimately, what I'd really like to do is wait until ABI=5 is mostly no longer being used then do a big refactor of this whole logic to simplify the code dramatically and then leave ABI=5 using the slow virtual function copying method for backwards-compatibility only.

Reply:
Sure, though the removal of sourceArray.loadData() and passing false to copyValues for compact should definitely be done imo. Why is this code branch a very rare case? I just happened to choose this change set to comment on, but it applies to the other one in this PR too. I can see a case of wanting to copy arbitrary (potentially single) indices from multiple source arrays to a single target array being common (e.g. point merge by groups) and think it's good that this is highlighted now as it does seem a fairly obvious limitation.

Reply:
It's a rare case, because for all registered types (which includes all types available in Houdini), this logic doesn't get invoked at all. It purely exists for the cases where you're registering and using an attribute type which isn't registered by default.

I'm going to do the aforementioned refactor now to produce a more tailored ABI=6 version. It's a bit of code duplication, but the ABI=6 one can be much simpler as a result of the introduction of AttributeArray::copyValues().

Reply:
Thanks for clarifying, that sounds good to me. Though is there any reason why rangeChecking couldn't be exposed as an argument on copyValuesUnsafe? Seems that that would resolve this too